### ISSUE TITLE
how to use Arabic language

### ISSUE THREAD

**Issue by mohammed-altmimi** [cnjinhao/nana#522]
how can use arabic language for a label or in text in nana 
like (("السلام عليكم ")) or (("مرحبا "))
in a label or simple text

**Comment by cnjinhao** [comment]
all string should be encoded as UTF-8.
```
label.caption(u8"مرحب ");
```

**Comment by mohammed-altmimi** [comment]
i try this method before
and never Success

//............................................................
#include<nana/gui/widgets/label.hpp>
#include <nana/gui.hpp>
#include <nana/gui/widgets/button.hpp>

int main()
{
using namespace nana;
form fm;
label lb(fm,rectangle{20,20,100,100},true);

lb.caption(u8"السلام عليكم");


fm.show();
exec();

}

//................................................................
and this screen show for my app 
https://ibb.co/0chmDtC

**Comment by cnjinhao** [comment]
I found an issue that label incorrectly deals with the RTL language(Arabic).
Thanks for your feedback.

**Comment by mohammed-altmimi** [comment]
no i do`nt think there is any issue with nana 
i think there is a header file "nana/internationalization.hpp"
is used for supporting languages format
i try to understand the example code combined with the header file but i could not ...
becuase   "  i18n.load("en.txt");"
for where i get or download en.txt and is there one for arabic language
and is there anthor example   

![bidi_example1](https://user-images.githubusercontent.com/62220324/77253405-0d614380-6c6b-11ea-94cf-1fd672d479ee.png)

and thank you for trying help me

**Comment by mohammed-altmimi** [comment]
i try textbox , button   and the same problem
the problem not for label only 

![text](https://user-images.githubusercontent.com/62220324/77256300-67b7cf80-6c7e-11ea-834d-9171a81d99aa.png)

**Comment by cnjinhao** [comment]
Hi, @mohammed-altmimi 
I have now fixed the issue of label for displaying Arabic language. I am waiting for your feedback. After your confirm, I will commit the fix.
```
//lb is a label
lb.format(true);
lb.caption(u8"<color=red>لسلام </><color=blue>عليكم </><color=black>السلام عليكم </>");
```
Now, it displays
![rtl](https://user-images.githubusercontent.com/1246418/77390490-1b30d900-6dd1-11ea-9adc-8540db4c588c.png)

Is the order of Arabic words right?

**Comment by mohammed-altmimi** [comment]
did not work
i use linux ubuntu 18.4
I think the OS has a role in how the text works in the ui

![nanaproj](https://user-images.githubusercontent.com/62220324/77415349-651abe80-6dd3-11ea-995f-ff88ae8181ba.png)

**Comment by cnjinhao** [comment]
Hi, I haven't committed the fix. I will commit it later. Please look at the screenshot, is it correct?

**Comment by cnjinhao** [comment]
Fixes have been checked into the `hotfix-1.7.3` branch. Please try it.

**Comment by mohammed-altmimi** [comment]
did not work 
i try every think 
your screen shot  is correct  text direction
![nana_hotfix](https://user-images.githubusercontent.com/62220324/77453842-d6c12f80-6e08-11ea-9fb9-56125deef5c8.png)
![nanahotfix3](https://user-images.githubusercontent.com/62220324/77453854-da54b680-6e08-11ea-8dc4-24a7e9e3ef49.png)
![nanahotfix4](https://user-images.githubusercontent.com/62220324/77453864-dde83d80-6e08-11ea-9e7d-7b3a4291a6e5.png)
![nanahotfix5](https://user-images.githubusercontent.com/62220324/77453870-dfb20100-6e08-11ea-9a4e-84ce9788b06e.png)

**Comment by cnjinhao** [comment]
Thanks for your feedback. It’s another issue on Linux, I will fix it

**Comment by cnjinhao** [comment]
Hi, @mohammed-altmimi 
Partial support of Arabic reshaping for Linux is added(support of ligatures not implemented). Please try the `hotfix-1.7.3` branch, I am looking forward to your feedback.

**Comment by mohammed-altmimi** [comment]
Yes, there is some progress
The direction of the text is correct .. But drawing letters is incorrect in some places of the word or sentence .. Some words I put at the end of the sentence become at the beginning of the sentence ..
...................................... side information .......... ......................
Each letter in the Arabic language is drawn in three ways, according to the position of the letter in the word
For example, the letter "ب"
Draw at the beginning of the word "بـ"
Draw in the middle of the word "ـبـ"
Draw at the end of the word "ـب"
.................................................. ...........................................
There is an issue in identifying the correct drawing of the letter in its current location of the word.

Thank you so much
 
![hotfix-45](https://user-images.githubusercontent.com/62220324/77748928-377f8080-7032-11ea-9480-6305e1e9ba48.png)

**Comment by cnjinhao** [comment]
Hi, @mohammed-altmimi 
A fix has been checked into `hotfix-1.7.3` branch. Please try it, I am looking for your feedback.

**Comment by mohammed-altmimi** [comment]
yes 👍👍
very gooood 
it`s work 
think you 
i will try other Widgets and give you feedback 
![fixed_nana](https://user-images.githubusercontent.com/62220324/78078786-8707d880-73b3-11ea-812c-c5c7147a8969.png)
